lib/rspec/core/source.rb

```diff
@@ -0,0 +1,70 @@
+module RSpec
+  module Core
+    # @private
+    # Represents a Ruby source file and provides access to AST and tokens.
+    class Source
+      attr_reader :source, :path
+
+      def self.from_file(path)
+        source = File.read(path)
+        new(source, path)
+      end
+
+      def initialize(source_string, path=nil)
+        @source = source_string
+        @path = path ? File.expand_path(path) : '(string)'
+      end
+
+      def lines
+        @lines ||= source.split("\n")
+      end
+
+      def ast
+        @ast ||= begin
+          require 'ripper'
+          RSpec::Support.require_rspec_core 'source/node'
```

Comment:
Re-requiring every time `ast` is called seems unfortunate.  I know ruby will short circuit but it still adds the overhead of going into ruby's `require` logic to be short-circuited.  I think it's fine if these requires (and `source/token`, below) are at the top of the file instead of being in the methods themselves.  I think we should just delay the loading of the entire `Source` sub-system as a unit rather than doing it on a file-by-file basis.